### PR TITLE
pAI holographic form returns + full pAI software format refactor

### PR DESCRIPTION
**Pull request by Ephemeralis**
This is something I've wanted to do for quite a while. The old pAI software format was a ridiculous mess of hard-coded Topic HTML links and assorted other legacy code tidbits, making adding new software an absolute fucking nightmare.

I've moved all software schemas over into a datum model with events for installing the software, drawing its menu (and results), and its point in the pAI sidebar. New software entities are modeled around the following schema, which should be pretty self-explanatory (and documented in comments).

```
/datum/pai/software
    var/name = "Prototype Software"
    var/description = "The basic software upon which all else is based."
    var/category = "" //for listing categories in pAI interface
    var/sid = "software" //marker used inside Topic, set to a single word based on the software
    var/hrefline //signal that the software uses an overwritten href line for right-side menu display (see action_hrefline)

    var/ram = 0 //amount of ram resource required by the pAI to use this

/datum/pai/software/proc/action_use(mob/user, var/args)
    //thrown when the user accesses the software via the side menu once they've purchased it
    return

/datum/pai/software/proc/action_installed(mob/user)
    //thrown when the user purchases the software for the first time
    //use this to add things like verbs, hud displays, etc
    return

/datum/pai/software/proc/action_menu(mob/user)
    //return the required HTML that will be shown in the menu for the user
    return

/datum/pai/software/proc/action_hrefline(mob/user)
    //populate with code that gives a custom href line format (for HUDs with On/Off displays etc, see medical_hud and security_hud)
    return


```

All existing pAI software has been converted to this new type (including the unfinished camera-jack code from like.. 6 years ago) and has been largely untouched outside of some alterations to allow it to work with the new model.

Holoform functions identically as it did in the previous yogs implementation. Literally no changes made outside of minor state changes to prevent weird shit from occurring.

Self-repair and ventcrawling functions exist in the code, but are not currently enabled at this time.

Also goof: pre-emptively suck my dick.
#### Changelog

:cl:
rscadd: pAI have discovered how to use their holographic emitters once more, without RAM cost.
rscdel: pAI require their masters permission explicitly set via the pAI card itself in order to use their holographic form at all.
bugfix: Massive numbers of bugs with existing pAI software suites have been fixed, including the inability to change radio settings from within the pAI interface. Refreshing is now consistent across all software types, and Remote Control functions properly with camera suites of all kinds.
tweak: RAM costs for some pAI software suites have been adjusted to better reflect their usefulness.
tweak: Some pAI software suites no longer use generic names, and are instead shipped to the pAI core repo with their long names intact.
tweak: Intrepid pAI personality cores have discovered how to resist the effects of external wiping, though not forever.
/:cl:

### CODE REVIEW

**Comment by george99g** [comment]
```
code/modules/mob/living/silicon/pai/pai_software.dm:44:error: SW.sid: undefined type: SW.sid
code/modules/mob/living/silicon/pai/pai_software.dm:45:error: SW.action_menu: undefined type: SW.action_menu
code/modules/mob/living/silicon/pai/pai.dm:109:error: /datum/pai/software: undefined type path
code/modules/mob/living/silicon/pai/pai.dm:109:error: /datum/pai/software: undefined type path
code/modules/mob/living/silicon/pai/pai.dm:111:error: S.sid: undefined type: S.sid
code/modules/mob/living/silicon/pai/pai.dm:111:error: S: undefined type: S
code/modules/mob/living/silicon/pai/pai_software.dm:200:error: SW.hrefline: undefined type: SW.hrefline
code/modules/mob/living/silicon/pai/pai_software.dm:201:error: SW.action_hrefline: undefined type: SW.action_hrefline
code/modules/mob/living/silicon/pai/pai_software.dm:203:error: SW.sid: undefined type: SW.sid
code/modules/mob/living/silicon/pai/pai_software.dm:203:error: SW.name: undefined type: SW.name
code/modules/mob/living/silicon/pai/pai_software.dm:219:error: S.ram: undefined type: S.ram
code/modules/mob/living/silicon/pai/pai_software.dm:219:error: S.ram: undefined type: S.ram
code/modules/mob/living/silicon/pai/pai_software.dm:219:error: S.ram: undefined type: S.ram
code/modules/mob/living/silicon/pai/pai_software.dm:221:error: S: undefined type: S
code/modules/mob/living/silicon/pai/pai_software.dm:224:error: S.sid: undefined type: S.sid
code/modules/mob/living/silicon/pai/pai_software.dm:224:error: S.name: undefined type: S.name
code/modules/mob/living/silicon/pai/pai_software.dm:224:error: S.ram: undefined type: S.ram
code/modules/mob/living/silicon/pai/pai_software.dm:113:error: SW.sid: undefined type: SW.sid
code/modules/mob/living/silicon/pai/pai_software.dm:114:error: SW.action_use: undefined type: SW.action_use
code/modules/mob/living/silicon/pai/pai_software.dm:123:error: T: undefined type: T
code/modules/mob/living/silicon/pai/pai_software.dm:123:error: /datum/pai/software: undefined type path
code/modules/mob/living/silicon/pai/pai_software.dm:124:error: T.ram: undefined type: T.ram
code/modules/mob/living/silicon/pai/pai_software.dm:127:error: T: undefined type: T
code/modules/mob/living/silicon/pai/pai_software.dm:128:error: T.action_installed: undefined type: T.action_installed
code/modules/mob/living/silicon/pai/pai_software.dm:132:error: T: undefined type: T
```

**Comment by Ephemeralis** [comment]
George is such a joker, making it look like I forgot to include the dme changes.

Haha, what a guy.

**Comment by george99g** [comment]
screenshot from 2016-09-25 22 08 33
<insert usual squash your commits bitching>

**Comment by Cruix** [comment]
Some bugs as it is:
- Remote control does not work and causes runtimes due to the pai software datum not having a set_machine() proc
- alien, chicken, and carp Pai holograms have no sprite
- In the software interface, the refresh button just makes the screen blank
- The Pai has a button to toggle its PDA ringer, but it does nothing.
- Neither the Pai nor the card holder can edit the Pai's radio settings
- The universal translator and security hud cannot be toggled on
- The button to change a Pai's directives has no checks for whether the card is being held, so you can keep the menu open and change the Pai's directives as many time as you want

Several of these are probably live on the server right now, but it would be nice if you fixed them too.

**Comment by Ephemeralis** [comment]
I'm slogging away at the radio bugs but am honestly far out of my depth. Can't figure out what's going on at the moment.

**Review by Ephemeralis — COMMENTED** [review]
To clarify, the changes to camera.dm allow pAI to use other aspects of their interface (such as other softwares) while the camera cycle interface is open at the same time. Hence the excessive checking.

**Comment by Kmc2000** [comment]
He's back alright

**Comment by Ephemeralis** [comment]
All bugs mentioned by Cruix are fixed.

File: code/game/objects/items/devices/paicard.dm [hunk 1]
```diff
@@ -55,11 +57,30 @@
...
+		return
+	else
+		return ..()
+
 /obj/item/device/paicard/Topic(href, href_list)
 
 	if(!usr || usr.stat)
 		return
 
+	if (!usr.contents.Find(src) || loc != usr)
```
Comment (Cruix) [inline code/game/objects/items/devices/paicard.dm]: Both of these checks do the same thing, and the second one is much cheaper.

Comment (Ephemeralis) [reply code/game/objects/items/devices/paicard.dm]: Fixed!

**Pull request merged**